README.md
# Surveybraid

A survey report writing system, building on surveyor and braid


NEWS
surveybraid 0.0-06 (2012-07-18)
---------------

New functionality:

* Includes methods for surveybraid as well as surveybraidppt


surveybraid 0.0-01 (2012-06-25)
---------------

First experimental release. Attempt to separate braid and survey functionality from the original surveyor package.

inst/tests/test-1-braid.R
# 
# Author: Andrie
#------------------------------------------------------------------------------


path <- tempdir()
latexPath <- file.path(path, "latex")
dir.create(latexPath, recursive=TRUE, showWarnings=FALSE)
graphPath <- file.path(latexPath, "graphics")
dir.create(graphPath, recursive=TRUE, showWarnings=FALSE)

sinkfile   <- "surveyor_test.tex"


q_data <- data.frame(
    Q1=c("Yes", "No", "Yes", "Yes"),
    Q4_1 = c(1, 2, 1, 2), 
    Q4_3=c(3, 4, 4, 3), 
    Q4_2=c(5, 5, 6, 6), 
    crossbreak=c("A", "A", "B", "B"), 
    crossbreak2=c("D", "E", "D", "E"),
    weight=c(0.9, 1.1, 0.8, 1.2)
)
varlabels(q_data) <- c(
    "Question 1", 
    "Question 4: red", "Question 4: blue", "Question 4: green", 
    "crossbreak",
    "crossbreak2",
    "weight")

q_data <- as.surveydata(q_data, renameVarlabels=TRUE)

names_cqrw <- c("cbreak", "question", "response", "weight")

sbraid <- as.braid(path = latexPath)

context("Test output to Latex")

test_that("working folders are empty", {
      if(!identical(list.files(graphPath), "")){
        lapply(list.files(graphPath), function(x)file.remove(file.path(graphPath, x)))
      }          
      if (file.exists(file.path(latexPath, sinkfile))){
        file.remove(file.path(latexPath, sinkfile))
      }
      expect_false(file.exists(file.path(graphPath, "fig0001.pdf")))
      expect_false(file.exists(file.path(graphPath, "fig0002.pdf")))
      
    })

test_that("surveyPlot works in Latex", {
      tbraid <- as.braid(path = latexPath, fileInner=sinkfile, outputType = "latex")
      s <- as.surveyor(q_data, q_data$crossbreak, q_data$weight, defaults=surveyorDefaults(printTable=FALSE))
      t <- as.surveybraid(s, tbraid)
      
      braidHeading(tbraid, "Test")
      surveyPlot(t, "Q1", statsBin, plotBar)
      surveyPlot(t, "Q4", statsBin, plotBar)
      braidSave(tbraid)
      
      expect_true(file.exists(file.path(latexPath, sinkfile)))
      expect_true(file.exists(file.path(graphPath, "Q1_.pdf")))
      expect_true(file.exists(file.path(graphPath, "Q4_.pdf")))
      
    })



R/surveybraid.R

#' Creates object of class surveybraid.
#' 
#' A surveybraid object consists of a surveyor and braid object.
#' 
#' @param surveyor surveyor object
#' @param braid braid object  
#' @return A list object of class \code{surveyorbraid}
#' @export
as.surveybraid <- function(surveyor, braid){
  if (!is.surveyor(surveyor)) stop("surveybraid: surveyor must be a surveyor object")
  if (!is.braid(braid))       stop("surveybraid: braid must be a braid object")
  structure(
      list(
          surveyor = surveyor,
          braid      = braid
      ), 
      class = "surveybraid"
  )
}

#' Tests that object is of class surveyorbraid.
#' 
#' @param x Object to be tested
#' @export
is.surveybraid <- function(x){
  if(!inherits(x, "surveybraid")) return(FALSE)
  if(is.surveyor(x$surveyor) & is.braid(x$braid)) TRUE else FALSE
}

#------------------------------------------------------------------------------

#' Codes and plots a survey question.
#' 
#' This is the top level function that determines how a question is processed,
#' coded, printed and plotted.
#' 
#' @param x surveybraid object
#' @param qid Question id
#' @param statsFunction  A surveyor stats function
#' @param plotFunction A surveyor plot function
#' @param codeFunction A surveyor stats function
#' @param onlyBreaks Numeric vector that limits crossbreak processing
#' @param outputType "latex", "ppt" or "device": Specifies destination of ouput
#' @param plotSize Size in inches of plot output, e.g. c(4,3)
#' @param plotMultiplierLimits Numeric vector of length two, indicating lower and upper limit of vertical plot resizing
#' @param addPlotTitle If TRUE, adds question text to plot title
#' @param ... Other parameters passed to \code{\link[surveyor]{surveyPlot}}
#' @method surveyPlot surveybraid
#' @export
#' @seealso \code{\link{as.surveyor}}
surveyPlot.surveybraid <- function(
    x,
    qid,
    statsFunction = "statsGuess",
    plotFunction = "plotGuess",
    codeFunction = "codeQuickArray",
    onlyBreaks=seq_along(x$surveyor$crossbreak),
    outputType = x$braid$outputType,
    plotSize = x$braid$defaultPlotSize,
    #plotMultiplierLimits = if(outputType=="ppt") c(0.8, 1.2) else c(0.8, 2.5),
    plotMultiplierLimits = c(0.8, 2.5),
    addPlotTitle = FALSE,
    ...
  ){
  
  surveyor <- x$surveyor
  braid    <- x$braid
  
  #browser()
  if(outputType=="ppt") if(!require(braidppt)) stop("Unable to load package braidppt")
  
  if(!exists(qid, surveyor$sdata) & is.null(which.q(surveyor$sdata, qid))){
    message(paste(qid,": Question not found.  Processing aborted"))
    return(NULL)
  }
  message(qid)
  plot_title <- qTextCommon(surveyor$sdata, qid)
  surveyor$plot_title <- plot_title
  if(outputType=="latex"){
    braidHeading(
        braid, 
        paste(qid, plot_title), 
        headinglevel= "section",
        pagebreak=FALSE)
  }

  if(outputType=="ppt"){
    braidpptNewSlide(braid, title=qid, subtitle=plot_title)
  }
  
  
  lh <- surveyPlot(surveyor, qid, statsFunction, plotFunction, codeFunction, 
      onlyBreaks=onlyBreaks, addPlotTitle=addPlotTitle, ...)
  
  
  lapply(seq_along(lh), function(i){
      catString <- surveybraidPrintQuestion(
          onlyBreaks[i],
          surveyor,
          braid,
          qid,
          lh[[i]],
          plotSize,
          outputType=outputType,
          plotMultiplierLimits=plotMultiplierLimits
      )
      if(!identical(catString, "")) braidWrite(braid, catString)
    }
  )
    	
  return(invisible(NULL))
}

#' Prints surveyor question. 
#' 
#' @inheritParams surveyPlot.surveybraid
#' @keywords internal
surveybraidPrintQuestion <- function(i, surveyor, braid, qid, h, plotSize, outputType, 
    plotMultiplierLimits=c(1, 1)){
  
  
  if(class(h$plot)=="text") return(h$plot)
  
  # Print plot
  filename <- paste(qid, "_", names(surveyor$crossbreak)[i], ".", braid$graphicFormat, sep="")
  message(paste(" --", filename))
  
  # Adjust vertical size of plot depending on number of questions
  # Make the assumption that 7 questions can fit on a plot
  # Limit vertical size to [1, 3]*size of default
  plotMin <- plotMultiplierLimits[1]
  plotMax <- plotMultiplierLimits[2]
  height_multiplier <- ifelse(
      is.numeric(h$nquestion), 
      min(plotMax, max(plotMin, h$nquestion / 7)),
      plotMin
  )
  switch(outputType,
      latex = {
        braidPlot(braid, h$plot, filename=filename,
            width=plotSize[1], height=(plotSize[2] * height_multiplier), Qid=qid)
      },
      ppt = {
        stopifnot(require(braidppt))
        braidppt::braidpptPlot(braid, h$plot, filename=filename,
            width=plotSize[1], height=(plotSize[2] * height_multiplier), Qid=qid)
      }
  )
  
  catString <- if(surveyor$defaults$printTable) tableGuess(h) else ""
  
  return(catString)
}





R/surveybraidppt.R

#' Creates object of class surveybraidppt.
#' 
#' A surveybraidppt object consists of a surveyor and braidppt object.
#' 
#' @param surveyor surveyor object
#' @param braid braid object  
#' @return A list object of class \code{surveyorbraid}
#' @export
as.surveybraidppt <- function(surveyor, braid){
  if (!is.surveyor(surveyor)) stop("surveybraidppt: surveyor must be a surveyor object")
  if (!is.braidppt(braid))       stop("surveybraidppt: braid must be a braid object")
  structure(
      list(
          surveyor = surveyor,
          braid      = braid
      ), 
      class = "surveybraidppt"
  )
}

#' Tests that object is of class surveyorbraid.
#' 
#' @param x Object to be tested
#' @export
is.surveybraidppt <- function(x){
  if(!inherits(x, "surveybraidppt")) return(FALSE)
  if(is.surveyor(x$surveyor) & is.braidppt(x$braid)) TRUE else FALSE
}

#------------------------------------------------------------------------------

#' Codes and plots a survey question.
#' 
#' This is the top level function that determines how a question is processed,
#' coded, printed and plotted.
#' 
#' @param x surveybraidppt object
#' @param qid Question id
#' @param statsFunction  A surveyor stats function
#' @param plotFunction A surveyor plot function
#' @param codeFunction A surveyor stats function
#' @param onlyBreaks Numeric vector that limits crossbreak processing
#' @param outputType "latex", "ppt" or "device": Specifies destination of ouput
#' @param plotSize Size in inches of plot output, e.g. c(4,3)
#' @param plotMultiplierLimits Numeric vector of length two, indicating lower and upper limit of vertical plot resizing
#' @param addPlotTitle If TRUE, adds question text to plot title
#' @param ... Other parameters passed to \code{\link[surveyor]{surveyPlot}}
#' @method surveyPlot surveybraidppt
#' @export
#' @seealso \code{\link{as.surveyor}}
surveyPlot.surveybraidppt <- function(
    x,
    qid,
    statsFunction = "statsGuess",
    plotFunction = "plotGuess",
    codeFunction = "codeQuickArray",
    onlyBreaks=seq_along(x$surveyor$crossbreak),
    outputType = x$braid$outputType,
    plotSize = x$braid$defaultPlotSize,
    #plotMultiplierLimits = if(outputType=="ppt") c(0.8, 1.2) else c(0.8, 2.5),
    plotMultiplierLimits = c(0.8, 1.2),
    addPlotTitle = TRUE,
    ...
  ){
  
  surveyor <- x$surveyor
  braid    <- x$braid
  
  #browser()
  if(!require(braidppt)) stop("Unable to load package braidppt")
  
  if(!exists(qid, surveyor$sdata) & is.null(which.q(surveyor$sdata, qid))){
    message(paste(qid,": Question not found.  Processing aborted"))
    return(NULL)
  }
  message(qid)
  
  lh <- surveyPlot(surveyor, qid, statsFunction, plotFunction, codeFunction, 
      onlyBreaks=onlyBreaks, addPlotTitle=addPlotTitle, ...)
  
  
  lapply(seq_along(lh), function(i){
      catString <- surveybraidpptPrintQuestion(
          onlyBreaks[i],
          surveyor,
          braid,
          qid,
          lh[[i]],
          plotSize,
          outputType=outputType,
          plotMultiplierLimits=plotMultiplierLimits
      )
      if(!identical(catString, "")) braidWrite(braid, catString)
    }
  )
    	
  return(invisible(NULL))
}

#' Prints surveyor question. 
#' 
#' @inheritParams surveyPlot.surveybraidppt
#' @keywords internal
surveybraidpptPrintQuestion <- function(i, surveyor, braid, qid, h, plotSize, outputType, 
    plotMultiplierLimits=c(1, 1)){
  
  stopifnot(require(braidppt))
  if(inherits(h$plot, "text")) return(h$plot)
  
  # Print plot
  filename <- paste(qid, "_", names(surveyor$crossbreak)[i], ".", braid$graphicFormat, sep="")
  message(paste(" --", filename))
  
  # Adjust vertical size of plot depending on number of questions
  # Make the assumption that 7 questions can fit on a plot
  # Limit vertical size to [1, 3]*size of default
  plotMin <- plotMultiplierLimits[1]
  plotMax <- plotMultiplierLimits[2]
  height_multiplier <- ifelse(
      is.numeric(h$nquestion), 
      min(plotMax, max(plotMin, h$nquestion / 7)),
      plotMin
  )

  plot_title <- qTextCommon(surveyor$sdata, qid)
  surveyor$plot_title <- plot_title
  
  braidpptNewSlide(braid, title=qid, text=plot_title)
  braidpptPlot(braid, h$plot, filename=filename,
      width=plotSize[1], height=(plotSize[2] * height_multiplier), Qid=qid)
  return(NULL)
}





inst/examples/example1.R
# Prepares folders for braidsurveyor output

latexPath <- file.path(tempdir(), "latex")
graphPath <- file.path(latexPath, "graphics")
dir.create(latexPath, recursive=TRUE, showWarnings=FALSE)
dir.create(graphPath, recursive=TRUE, showWarnings=FALSE)

# Defines data frame with survey data

qData <- data.frame(
    Q1=c("Yes", "No", "Yes", "Yes"),
    Q4_1 = c(1, 2, 1, 2), 
    Q4_3=c(3, 4, 4, 3), 
    Q4_2=c(5, 5, 6, 6), 
    crossbreak=c("A", "A", "B", "B"), 
    crossbreak2=c("D", "E", "D", "E"),
    weight=c(0.9, 1.1, 0.8, 1.2)
)
varlabels(qData) <- c(
    "Question 1", 
    "Question 4: red", "Question 4: blue", "Question 4: green", 
    "crossbreak",
    "crossbreak2",
    "weight")

# Creates surveydata object

qData <- as.surveydata(qData, renameVarlabels=TRUE)

# Creates surveyor and surveybraid object

s <- as.surveyor(qData, crossbreak=qData$crossbreak, weight=qData$weight, 
    defaults=surveyorDefaults(printTable=FALSE))
b <- as.braid(path = latexPath, fileInner="surveyor_test.tex", outputType = "latex")
sb <- as.surveybraid(s, b)

# Write heading and two plots; then save braid

braidHeading(b, "Test")
surveyPlot(sb, "Q1", statsBin, plotBar)
surveyPlot(sb, "Q4", statsBin, plotBar)
braidSave(b)

# Check that braid output exists

file.exists(file.path(latexPath, "surveyor_test.tex"))
file.exists(file.path(graphPath, "Q1_.pdf"))
file.exists(file.path(graphPath, "Q4_.pdf"))
      

